### DISCUSSION TITLE
Add support for the post-quantum ML-KEM KE algorithm in openssl plugin

### DISCUSSION THREAD

**Discussion by geedo0** [strongswan/strongswan#2175] [Ideas]
Hi there, I just wanted to open a discussion here to get feedback on this idea before raising it to a feature request. Please let me know if anything is unclear or needs more detail

**Is your feature request related to a problem? Please describe.**

Post-Quantum (PQ) cryptography aims to mitigate the threat quantum computing poses to confidentiality of our data in transit. In mid-2024, NIST is expected to standardize the PQ algorithm ML-KEM (formerly known as Kyber). An [IETF draft](https://datatracker.ietf.org/doc/draft-kampanakis-ml-kem-ikev2/) has been submitted for registering this algorithm in IANA's IKEv2 parameter registry. We would like to contribute an additional implementation of this algorithm in the openssl plugin.

Currently, strongSwan-6 supports this algorithm in the oqs plugin by using the Open Quantum Safe library `liboqs` directly. This library comes with a strong [disclaimer](https://github.com/open-quantum-safe/liboqs?tab=readme-ov-file#limitations-and-security) around usage in a production. While there are [talks](https://www.linuxfoundation.org/press/announcing-the-post-quantum-cryptography-alliance-pqca) to transition the OQS project to something more production ready with its recent membership in the Linux Foundation, it's still unclear what exactly that means (new library, supporting PQ in existing work, or improving liboqs itself). Some signs signal towards continued development of OQS's [OpenSSL-3 provider](https://github.com/open-quantum-safe/oqs-provider) which uses the OpenSSL KEM API.

This feature request proposes using [AWS-LC](https://github.com/aws/aws-lc) (an OpenSSL derivative) as a means to provide PQ confidentiality while providing a path toward future PQ algorithm implementations in upstream OpenSSL.

**Describe the solution you'd like**

The openssl plugin can be updated ([here](https://github.com/strongswan/strongswan/blob/7859f5ddd8a9b73d69c8911003a1e32e7e8b7797/src/libstrongswan/plugins/openssl/openssl_plugin.c#L374)) to conditionally provide support for the 3 different [Kyber algorithms](https://github.com/strongswan/strongswan/blob/0555bdd46b2c5aec301c8ef332ff5d3d2375e716/src/libstrongswan/crypto/key_exchange.h#L71-L74) KE's. It will register an implementation which will be able to handle arbitrary KEM algorithms.

AWS-LC currently provides implementations for Kyber Round 3 with plans to add support for the NIST standardized ML-KEM algorithms as soon as they are available. This is exposed via the [EVP KEM API](https://github.com/awslabs/aws-lc/blob/7f2e21ba685b41fda1c384378ada2769a3929381/include/openssl/evp.h#L713-L766) which was introduced in [OpenSSL-3](https://www.openssl.org/docs/man3.1/man3/EVP_PKEY_encapsulate.html). Implementations for these KE's can be implemented in a new source file `openssl_kem.c` which would implement all KEMs using the KEM API.

**Describe alternatives you've considered**

Alternative 1: Create a stand-alone AWS-LC plugin. This plugin would offer very limited functionality on its own (just the ML-KEM KEs and eventually ML-DSA signatures). Our hope is to encourage re-use with upstream OpenSSL when it provides support for ML-KEM. We also don't anticipate to add significant functionality that is unique to AWS-LC so this plugin would have limited utility.

Alternative 2: Add the implementation to the oqs plugin. The oqs plugin happens to link in libcrypto and can be coaxed into calling the libcrypto APIs for PQ implementations. This is messy and mixes concerns without much benefit.

**Additional context**

While the AWS-LC KEM API is similar to OpenSSL-3's API, it is not identical. The only KEM implemented by OpenSSL using this API is [RSA](https://github.com/openssl/openssl/pull/12750). Until the upstream OpenSSL implementation of ML-KEM is available, there remains a possibility of API inconsistencies that need to be ironed out. This is because it is their policy not to introduce non-standardized algorithms in their library. Long-term, we would like to be drop-in compatible, but we are waiting for the final shape before committing to an interface.

We are in a transition period between two versions of the same algorithm "Kyber Round 3" and "NIST ML-KEM". The naming of the constants in `key_exchange.h` and their IANA code points will change. A strategy needs to be proposed (possibly in a separate thread) to address questions around deprecation and migration across both the openssl and oqs plugins.

BoringSSL (AWS-LC's parent and descendent of OpenSSL) has yet [another interface](https://github.com/google/boringssl/blob/master/include/openssl/experimental/kyber.h) for Kyber which is incompatible with the OpenSSL-3's KEM API. We offered to upstream the KEM API implementation, but they are unwilling to expand the EVP_PKEY interface in this way so it's unlikely that BoringSSL will use the same KEM interface.

This feature request focuses on confidentiality of data-in-transit using ML-KEM. It does not address the PQ threat against authenticity which requires PQ signatures such ML-DSA (formerly known as Dilithium). This is because data-in-transit is a current threat due to store-now-harvest later attacks. Exploiting authenticity would require an attacker possessing a quantum computer during connection initiation. AWS-LC does support Dilithium, but it is not included in the default build.

**Comment by tobiasbrunner** [comment]
Since we are waiting for identifiers for ML-KEM as well, before merging the PQC changes in the _six-beta_ branch, I don't really see any major issues. Providing KEMs via plugins for third-party libraries is planned anyway once these libraries provide implementations (and stable APIs I guess, just to avoid lots of ifdefs, it's not really an issue if there are multiple APIs as long as this is hidden in the plugins).

**Comment by geedo0** [comment]
Great, I'll continue to get more information on these competing APIs and figure out the most stable path. While it can all be abstracted at the plugin layer, one of our goals is to minimize these kinds of compatibility issues.

I also heard that the IETF will be talking about ML-KEM identifiers this week so there may be news very soon on that.

Do you have any guidance on when the six-beta branch gets merged? We're not in a rush, so if we can avoid ifdefs and checking in less than stable APIs by leveraging time/information that seems like the best option in the long-term.

**Reply by tobiasbrunner** [reply]
> Do you have any guidance on when the six-beta branch gets merged?

I guess parts of it I'll merge relatively soon after IKEv2 identifiers for ML-KEM have been (pre-)allocated. Many changes are not technically dependent on it. For instance, multiple key exchanges also work with classic (EC)DH methods and the rekey collision refactoring is only in that branch because the code it depends on changed a lot. But we wanted these changes to be part of the next major release, so merging the branch to master means that the next release will be 6.0 and it would make sense to have at least one standardized KEM to go with it.  However, that doesn't mean we already need to have all plugins ready by then. There will definitely be some time after the merge to get the multi-KE changes tested in the wild a bit more (we also have to sort out some stuff with changing the default plugin selection). I also hope FrodoKEM gets an identifier (via ISO and then IETF) before we eventually make the release as we have a plugin for it that doesn't depend on any third-party libraries.

**Comment by geedo0** [comment]
Found one issue while trying to get this to work. [This](https://github.com/strongswan/strongswan/blob/six-beta/src/libstrongswan/crypto/key_exchange.h#L152) `set_seed` method used by the unit tests can't be implemented in AWS-LC because we do not support overriding the RNG through the public interface like OQS does. The comments suggest it's optional, but the tests won't pass without it being implemented. The KATs for all algorithms are run in our own testing which does the necessary RNG manipulation through private interfaces. I could do something similar to the DH KE's where a hardcoded key is set, but that requires changing how all the KEM tests are coded up. I'm not sure how common this kind of limitation is across all libraries so I can't say for sure if you'll run into this issue when implementing KEMs in other plugins. Any thoughts?

**Reply by tobiasbrunner** [reply]
I see. That's unfortunate, but could definitely be an issue with other libraries as well. OpenSSL, I suppose, as `EVP_PKEY_encapsulate()` will just allocate the shared secret from some internal RNG (setting the private key from a static value might be possible via `OSSL_PKEY_PARAM*`, similar to what we do for DH, if they extend that for KEMs).

I saw that Botan provides a [KEM API](https://botan.randombit.net/handbook/api_ref/ffi.html#public-key-encapsulation) where the shared secret is allocated with `botan_pk_op_kem_encrypt_create_shared_key()` that takes a `botan_rng_t` instance (similar when generating the private key). Such RNGs can be customized via `botan_rng_init_custom()`. So for tests we could probably provide an instance that reads data from the DRBG passed to `set_seed()`.

And wolfSSL provides an [experimental Kyber API](https://github.com/wolfSSL/wolfssl/blob/master/wolfssl/wolfcrypt/kyber.h) that has `wc_KyberKey_MakeKeyWithRandom()` and `wc_KyberKey_EncapsulateWithRandom()` functions that directly take the random data. The non-`WithRandom` prototypes take a `WC_RNG` instance, which doesn't seem easily customizable (at least not just for the tests). Now, for the former they completely ignore the random data if wolfSSL uses liboqs (or pqm4) as a backend, because they claim it's not possible to pass custom random data.  I guess that's true in the sense that there is no API similar to `*WithRandom` that directly takes the random data. On the other hand, for the `WC_RNG` versions, they do something similar to what we do, that is, wrapping existing RNGs in a callback registered via `OQS_randombytes_custom_algorithm()`. They seem to use a global mutex around calls that read from the RNG, we use thread-local values. And because we have a `drbg_t` wrapper around pre-defined entropy, our _oqs_ plugin also works with the test vectors. Anyway, with wolfSSL we could run the tests via `*WithRandom()` if the internal Kyber implementation is used (that would be the case in our CI builds). And to avoid calling two different sets of functions in test and regular scenarios, we'd have to always call these functions and allocate the random data outside if not reading it from the DRBG set via `set_seed()`.

Not providing `set_seed()` currently has two problems. First, it's not actually optional at the moment. While it's documented in the header as such, the crypto tester will just call it anyway and cause a crash. You could provide a dummy implementation that does nothing (e.g. `return_true()`), but this will eventually cause the test vectors to fail anyway. Now the problem with the latter is that [algorithms can be disabled](https://docs.strongswan.org/docs/5.9/plugins/test-vectors.html) at runtime based on failing those tests (for our CI it's obviously not ideal either). So if we actually make the method optional, I'm not sure what to do with such implementations. Maybe this needs a new `crypto_test` option that causes KEM implementations that are skipped due to a missing `set_seed()` to get accepted.

**Reply by geedo0** [reply]
> ...OpenSSL, I suppose, as EVP_PKEY_encapsulate() will just allocate the shared secret from some internal RNG (setting the private key from a static value might be possible via OSSL_PKEY_PARAM*, similar to what we do for DH, if they extend that for KEMs).

Yes, that will be the case for OpenSSL. I think this is why the Kyber authors specified the KATs with a seed and the specific DRBG algorithm. This would've been the only way to make the whole operation deterministic. We can construct an EVP_PKEY and related CTX from raw bytes, but we can't force the shared secret value. I took a look at the OpenSSL RNG's man page and they don't make it simple (and explicitly discourage) to override the library's RNG ([EVP_RAND](https://www.openssl.org/docs/manmaster/man7/EVP_RAND.html)). I agree, that it'll be interesting to see what OSSL_PKEY_PARAMs upstream OpenSSL decides to support for PQ-KEMs. The OQS Provider only uses this [interface](https://github.com/open-quantum-safe/oqs-provider/tree/main) to surface the keys of a hybrid contruction. Notably, they don't support overriding the RNG through the OpenSSL Provider interface.

To complement your investigation of other libraries, I can confirm that BoringSSL also doesn't support RNG overrides in the public interface. It has a private interface for testing KATs [here](https://github.com/google/boringssl/blob/master/crypto/kyber/internal.h). Similarly, we have [this](https://github.com/aws/aws-lc/blob/main/crypto/rand_extra/pq_custom_randombytes.h) interface in AWS-LC which was created for a similar usecase of passing KATs in third party software, but it's really hacky and is planned for removal.

> First, it's not actually optional at the moment. While it's documented in the header as such, the crypto tester will just call it anyway and cause a crash. You could provide a dummy implementation that does nothing (e.g. return_true()), but this will eventually cause the test vectors to fail anyway

Would it be possible to downgrade this test to a warning for certain libraries?

> Now the problem with the latter is that algorithms can be disabled at runtime based on failing those tests (for our CI it's obviously not ideal either). So if we actually make the method optional, I'm not sure what to do with such implementations. Maybe this needs a new crypto_test option that causes KEM implementations that are skipped due to a missing set_seed() to get accepted.

How open is the project to such a workaround? From my perspective, I wouldn't mind losing this automatic self-test and unload functionality for AWS-LC. I am concerned about how this would be handled with upstream OpenSSL, that seems less acceptable.

**Reply by geedo0** [reply]
It looks likely that we will be adding "derandomized" versions of the keygen and encapsulate APIs for ML-KEM in AWS-LC to support this use case and other efforts. Basically, these versions of the APIs will accept a buffer containing the 32-bytes of randomness needed for each of these operations. This means we can support strongSwan's testing by calling into the DRBG passed in by the test_vector plugin and passing the results into the derandomized APIs. Does that work?

I do suggest simplifying strongSwan's KEM vectors to just specify the random bytes to use instead of accepting a seed for a DRBG. The NIST CAVP vectors simply specify the 32-bytes of entropy [directly](https://github.com/usnistgov/ACVP-Server/blob/master/gen-val/json-files/ML-KEM-keyGen-FIPS203/internalProjection.json#L16) which avoids the indirection (z is defined in the [standard](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.203.ipd.pdf)).

**Reply by tobiasbrunner** [reply]
> Does that work?

That sounds similar to what wolfSSL provides, right? (i.e. something like `wc_KyberKey_MakeKeyWithRandom()`.) If so, that should work. But as mentioned above, it's not ideal to call different API functions during the tests and in regular use, so we'd probably always use that  API and allocate randomness via an `rng_t/drbg_t` outside even when not testing.

> I do suggest simplifying strongSwan's KEM vectors to just specify the random bytes to use instead of accepting a seed for a DRBG. The NIST CAVP vectors simply specify the 32-bytes of entropy [directly](https://github.com/usnistgov/ACVP-Server/blob/master/gen-val/json-files/ML-KEM-keyGen-FIPS203/internalProjection.json#L16) which avoids the indirection (z is defined in the [standard](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.203.ipd.pdf)).

We need test vectors not for single operations (which the CAVP vectors seem to provide) but for the complete key exchange. So what's called "seed" in our test vectors has to be the source for all the secrets (for both ends of the exchange, not just `z` for the initiator).

For classic (EC)DH the seed is just the secret key for both ends that gets split in half before passing one halve to each instance. And with the KATs during the submission rounds, the seeded DRBG provided all the secrets when called in the right order (note that the same DRBG instance is passed to both KEM instances). The latter also worked for all algorithms the same way (which is particularly important for the _oqs_ plugin that wraps multiple of them).

But we could theoretically work with static entropy values by providing a pseudo-DRBG that just hands it out as is (I feel like we had something like this at some point). So the seed could then just be the concatenated `z` and `m` values. If we continue to support HQC, we have to change the DRBG allocation anyway, so it might be possible to support such "static" vectors too (i.e. let the test vector specify the DRBG type and provide something like `DRBG_STATIC` for such vectors and the others could continue to select `DRBG_CTR_AES256` or something else, like `SHAKE256` for HQC, although that's currently not a DRBG type).

**Reply by geedo0** [reply]
> That sounds similar to what wolfSSL provides, right? (i.e. something like wc_KyberKey_MakeKeyWithRandom().) If so, that should work. But as mentioned above, it's not ideal to call different API functions during the tests and in regular use, so we'd probably always use that API and allocate randomness via an rng_t/drbg_t outside even when not testing.

Yes, that's exactly what it would look like and I agree that that's the most consistent approach here. Any specific preference for the "default" DRBG to use for the plugin? Otherwise, I'll see what the best option is that complies with the NIST standard.

> We need test vectors not for single operations (which the CAVP vectors seem to provide) but for the complete key exchange. So what's called "seed" in our test vectors has to be the source for all the secrets (for both ends of the exchange, not just z for the initiator).

Ahh, that explains the disconnect between CAVP and the OQS/Crystals vectors. The approach you have now definitely works for your requirements and is easy enough to work with. Like you and your the concerns around HQC, I'm worried how well this generalizes across different algorithms and KAT schemes.

**Reply by tobiasbrunner** [reply]
> Any specific preference for the "default" DRBG to use for the plugin?

It does not have to be a DRBG, we can just use an `rng_t` with `RNG_STRONG` (which is often backed by a DRBG anyway).  I was just thinking that using a `drbg_t` even when not testing might simplify the code. But we could also just check if a `drbg_t` was set via `set_seed()` and then either get the 32 bytes from there or create an `rng_t` and use that (which would be necessary to seed a `drbg_t` anyway, so not sure if this would really be a simplification).

> I'm worried how well this generalizes across different algorithms and KAT schemes

Yeah, we'll have to see. It wouldn't be that much of a problem to add a new field to `ke_test_vector_t` that specifies the DRBG type and if not set default to `DRBG_CTR_AES256` so we wouldn't have to change any of the existing vectors.

**Comment by geedo0** [comment]
Just to provide an update before this thread runs dormant for awhile. I have this working in a private branch with the `set_seed` method stubbed out. I'll hold off on submitting a PR until we have both the final ML-KEM implementation available and the derandomized versions of the APIs in place. That way, the tests will pass and we can interoperate with the OQS plugin. I'll summarize everything we discussed so far and convert this to a proper Github Issue later this week.
